src/connection-loop.ts

```diff
@@ -0,0 +1,229 @@
+import {resolver} from './resolver.js';
+import {sleep} from './sleep.js';
+
+const DEBOUNCE_DELAY = 10;
+
+const MIN_DELAY = 30;
+const MAX_DELAY = 60_000;
+
+const MAX_CONNECTIONS = 3;
+
+type Duration = {duration: number; ok: boolean};
+type Durations = Duration[];
+
+export interface ConnectionLoopDelegate {
+  closed(): boolean;
+  invokeExecute(): unknown;
+  debounceDelay?: number;
+  maxConnections?: number;
+  printTime(): string;
+  watchdogTimer?: number;
+}
+
+export class ConnectionLoop {
+  private _pendingResolver = resolver<void>();
+  private _currentResolver = resolver<void>();
+  private _pushCounter = 0;
+
+  private readonly _delegate: ConnectionLoopDelegate;
+  private _lastSendTime = 0;
+
+  constructor(delegate: ConnectionLoopDelegate) {
+    this._delegate = delegate;
+    this.run();
+  }
+
+  async execute(): Promise<void> {
+    this._pendingResolver.resolve();
+    // await a promise to allow the currentPushResolver to be created inside the
+    // connection loop.
+    await Promise.resolve(0);
+    await this._currentResolver.promise;
+  }
+
+  async run(): Promise<void> {
+    const durations: Durations = [];
+    let delay = MIN_DELAY;
+    let lastConnectionErrored = false;
+    const delegate = this._delegate;
+    const {
+      debounceDelay = DEBOUNCE_DELAY,
+      maxConnections = MAX_CONNECTIONS,
+      watchdogTimer,
+    } = delegate;
+
+    console.log('Starting connection loop', delegate.printTime());
+
+    while (!delegate.closed()) {
+      console.log('Waiting for a push', delegate.printTime());
+
+      if (watchdogTimer) {
+        await Promise.race([
+          this._pendingResolver.promise,
+          sleep(watchdogTimer),
+        ]);
+      } else {
+        await this._pendingResolver.promise;
+      }
+
+      // This resolvers is used to make the individual calls to push have the
+      // correct "duration".
+      const currentResolver = resolver();
+      this._currentResolver = currentResolver;
+      // Don't let this rejected promise escape.
+      currentResolver.promise.catch(() => 0);
+
+      if (debounceDelay) {
+        await sleep(debounceDelay);
+      }
+
+      // This resolver is used to wait for incoming push calls.
+      this._pendingResolver = resolver();
+
+      if (lastConnectionErrored || this._pushCounter >= maxConnections) {
```

Comment:
The first clause of this `or` doesn't seem right to me. We don't want to `waitUntilAvailableConnection()` for a single connection error.

Reply:
The intent was to ensure we only do one request at a time when the last request failed.

Changed it to:

```ts
if (this._counter >= (lastConnectionErrored ? 1 : maxConnections)) {
```

Reply:
This is needed. If not we end up with concurrent request when the last one errored. I verified this with making the change and looking at the test output.